Emulator framework and example I2C emulator

sjg20 opened the pull request:
This has two commits which attempt to implement an I2C emulator setup, where I2C devices can be emulated in software by emulators.

These drivers are used to emulate hardware devices, typically in native_posix, to support testing of various subsystems. For example, it is possible to write an emulator for an I2C compass such that it appears on the I2C bus and can be used just like a real hardware device.

Emulators often implement special features for testing. For example a compass may support returning bogus data if the I2C bus speed is too high, or may return invalid measurements if calibration has not yet been completed. This allows for testing that high-level code can handle these situations correctly. Test coverage can therefore approach 100% if all failure conditions are emulated.

sjg20 committed:
native_posix: Use the I2C emulator for the EEPROM test too

At present this test uses an EEPROM emulator. Reuse the same test to
also use the real Atmel AT2x driver and an AT24 emulator, via the I2C
emulation controller.

EEPROM reads and writes for eeprom1 go through the AT2x driver, which
converts them to I2C transactions, which are passed through to the
AT24 emulator for processing. This approach makes more use of 'real'
code, in this case the Atmel AT2x driver.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Simon Glass <sjg@chromium.org>

aescolar commented on drivers/emul/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	help
+	  Enable Emulation Driver Configuration
+	  These drivers are used to emulate hardware devices, typically in
+	  native_posix, to support testing of various subsystems. For example,
```
> typically in native_posix

Well.. only if that is your preferred debugging platform.
I mean, this works with any target with enough RAM. So I would not mention native_posix explicitly here, as you are not using any of its goodies or suffering any of its limitations.

sjg20 replied:
OK done

aescolar reviewed (commented):
Just a silly comment after a very quick browse thru the code.
I let those involved with I2C review, as there is nothing here specific to native_posix or the posix arch.

pabigot commented:
I'd like to step back to get a better understanding of the goals and constraints underlying these simulators.

#26484 seems pretty clear: it provides a simulated GPIO device that can pass the standard Zephyr test for GPIO devices, and so could be used to test a Zephyr component that uses those GPIOs by controlling the GPIO events and states observed and set by that component.

I understood the goal here to be to provide a I2C controller that would allow testing of a *driver* for an I2C device without having an instance of the device connected to a real I2C bus.  Specifically a developer could use a hardware description like:

```

gpio0: gpio@0 {
  compatible = "zephyr,sim-gpio";
  reg = <0>;
  status = "okay";
  ngpios = <32>;
};

i2c0: i2c@0 {
  compatible = "zephyr,sim-i2c";
  reg = <0>;
  status = "okay";
  eeprom: eeprom@56 {
    compatible = "atmel,at24";
    reg = <0x56>;
    label = "EEPROM";
    size = <1024>;
    pagesize = <16>;
    address-width = <8>;
    timeout = <5>;
  };
  hts221: hts221@5f {
    compatible = "st,hts221";
    reg = <0x5f>;
    label = "HTS221";
    drdy-gpios = <&gpio0 24 GPIO_ACTIVE_HIGH>;
  };
};
```

and produce an application that uses a simulated I2C controller driver with a simulated GPIO controller driver to run the real HTS221 sensor driver without a physical HTS221, and the real EEPROM driver without a physical Atmel AT24C08A.

If that's not the goal, I think we should start with an RFC issue that explores the goal in more detail.  (That might be a good idea anyway.)

If that is the goal, I'd really like to see a PR that's more focused.  This version includes a reference to a sensor Zephyr doesn't support.  It doesn't include a driver, but it does include something that emulates a driver.  I'm struggling to see what this enables.

What I come up with is that we could test an *application* (that we don't have), e.g. one that needs an HMC5883L magnetometer, and the goal is to see whether that application reacts properly to anomalous behavior injected by an emulated magnetometer driver.

I can see some value in that.  But I can see **much more** value if we're using a real driver (that would work with real hardware) where the behavior is controlled by emulating I2C bus or GPIO behavior.  First because then we're testing code that Zephyr provides, not an external application, and second because it easily extends to testing a temperature sensor, accelerometer, or any other device that interacts with an I2C bus.

Does that make sense?  Can that be the goal?

If so, the issues with devicetree will probably go away, because it's straightforward to have the `zephyr,sim-i2c` bus device provide driver-specific extensions that associate function tables (init, transfer) with device addresses, where the functions are provided by code that simulates specific bus (or gpio) behavior for the device.  I don't think the devicetree needs anything other than the right compatibles for the I2C and GPIO nodes to switch between hardware and emulation.  There need not be any Zephyr devices at all associated with the emulation code; the `emul` directory goes away (unless it moves to "samples" and provides core init/transfer functions that try to implement the observable interactions for a functioning device).

Making that work for the `atmel,at24` driver should be relatively easy:
* provide an init function that uses a 1 KiBy RAM buffer initialized to all 0xFF
* provide a transfer function that takes standard AT24 commands to mutate the content of the buffer
* add a module to `tests/drivers/eeprom` that on `native_posix` uses an overlay like the one above and hooks in the emulator functions

At that point `tests/drivers/eeprom` should pass on `native_posix`

That would be a far more compelling example than simulating a device that doesn't have a driver.

As a separate concern I think the reference to "native posix" may be due to architectural re-use of concepts from the [U-Boot Sandbox](https://github.com/u-boot/u-boot/blob/master/doc/arch/sandbox.rst) where the bootloader runs in a POSIX environment.  If this architecture assumes the ability to read devicetree content to identify dynamic libraries and files that need to be linked and mapped to set things up before the application gets started I'm not sure that can be done from Zephyr's native_posix environment.  I'm guessing that part of the role of the extra material in devicetree is to automate the linking of the emulation function table into the test application.  Perhaps that could be done, but we can get pretty far towards value by starting with a manual solution.

At any rate, that's where I am on this.  If the above is way off base, then please let's start with an RFC issue describing the problem and identifying stakeholder expectations before we dive into solutions.

sjg20 commented:
> I'd like to step back to get a better understanding of the goals and constraints underlying these simulators.
> 
> #26484 seems pretty clear: it provides a simulated GPIO device that can pass the standard Zephyr test for GPIO devices, and so could be used to test a Zephyr component that uses those GPIOs by controlling the GPIO events and states observed and set by that component.
> 
> I understood the goal here to be to provide a I2C controller that would allow testing of a _driver_ for an I2C device without having an instance of the device connected to a real I2C bus. Specifically a developer could use a hardware description like:

Yes that's right.

> If that is the goal, I'd really like to see a PR that's more focused. This version includes a reference to a sensor Zephyr doesn't support. It doesn't include a driver, but it does include something that emulates a driver. I'm struggling to see what this enables.

OK. I can move this over to the Atmel AT24.

> If so, the issues with devicetree will probably go away, because it's straightforward to have the `zephyr,sim-i2c` bus device provide driver-specific extensions that associate function tables (init, transfer) with device addresses, where the functions are provided by code that simulates specific bus (or gpio) behavior for the device. I don't think the devicetree needs anything other than the right compatibles for the I2C and GPIO nodes to switch between hardware and emulation. There need not be any Zephyr devices at all associated with the emulation code; the `emul` directory goes away (unless it moves to "samples" and provides core init/transfer functions that try to implement the observable interactions for a functioning device).

I'm not quite sure what you are saying here. We may end up with 50 emulators for different devices. Are you wanting those all to be in the zephyr,sim-i2c driver? I am trying to create a way to connect up the emulators to the drivers using the device tree.

> 
> Making that work for the `atmel,at24` driver should be relatively easy:
> 
> * provide an init function that uses a 1 KiBy RAM buffer initialized to all 0xFF
> * provide a transfer function that takes standard AT24 commands to mutate the content of the buffer

You still mean emulate at the I2C layer, right?

> * add a module to `tests/drivers/eeprom` that on `native_posix` uses an overlay like the one above and hooks in the emulator functions

How does it hook in the emulator functions?

> 
> At that point `tests/drivers/eeprom` should pass on `native_posix`
> 
> That would be a far more compelling example than simulating a device that doesn't have a driver.
> 
> As a separate concern I think the reference to "native posix" may be due to architectural re-use of concepts from the [U-Boot Sandbox](https://github.com/u-boot/u-boot/blob/master/doc/arch/sandbox.rst) where the bootloader runs in a POSIX environment. If this architecture assumes the ability to read devicetree content to identify dynamic libraries and files that need to be linked and mapped to set things up before the application gets started I'm not sure that can be done from Zephyr's native_posix environment. I'm guessing that part of the role of the extra material in devicetree is to automate the linking of the emulation function table into the test application. Perhaps that could be done, but we can get pretty far towards value by starting with a manual solution.

I actually worry that native_posix is not quite what we want, since people are always talking about connecting 'real' hardware into Zephyr. E.g. the i2c driver might talk to the SMBUS on the PC. That is quite different from the test-focused effort I am trying to start here. But for now, it seems to work.

> 
> At any rate, that's where I am on this. If the above is way off base, then please let's start with an RFC issue describing the problem and identifying stakeholder expectations before we dive into solutions.

I think this is fine except for the bit about not having an emul directory, etc. Will await your thoughts on that.

pabigot commented:
> ... 50 emulators for different devices.  Are you wanting those all to be in the zephyr,sim-i2c driver?

No, I'm saying the `zephyr,sim-i2c` driver should be a trivial intermediary, and that the logic and data associated with different emulated things can be injected at startup.  (I'm also saying they're not devices in the Zephyr sense, but hopefully that'll be more clear at the end of this.)

> I am trying to create a way to connect up the emulators to the drivers using the device tree.

I think that's the problem.  For basic functionality we don't need the linkage to be automated, and IMO it's premature to try to do it via devicetree (except that, after having developed a more detailed concept, what I'm suggesting below seems to come pretty close).

> You still mean emulate at the I2C layer, right?

Yes.

> How does it hook in the emulator functions?

Do something like `<gpio/gpio_sim.h>` in #26484 but for an intermediate layer like a driver rather than a leave node like gpio.

Add a new header `<i2c/i2c_sim.h>` that has something much like your `<drivers/emul/i2c_emul.h>`:

```
typedef int (*i2c_sim_init_t)(void *instance);
typedef int (*i2c_sim_transfer_t)(void *instance,
                                  struct i2c_msg *msgs, int num_msgs,
                                  int addr);
/* Functions that the i2c controller intermediary forwards to. */
struct i2c_sim_descriptor {
  i2c_sim_init_t init;
  i2c_sim_transfer transfer;
};

/* Node in a linked list of registered emulated i2c devices. */
struct i2c_sim_registration {
  sys_slist_t node;
  /* Device-specific state */
  void *instance;
  /* The function table used for this device. */
  const struct i2c_sim_descriptor *desc;
  /* I2C address at which device is reachable. */
  uint16_t addr;
};

/* Register an emulated device on the controller. */
int i2c_sim_register(struct device *dev,
		     const i2c_sim_descriptor *desc);
```

The implementation module for emulated I2C devices would provide a global symbol with a known name derived from the device compatible, like:

```
/* Pass i2c_sim instance into the module. */
int emul_atmel_at24_register(struct device *dev);
```

The `emul_atmel_at24` implementation module would use `DT_INST_FOREACH_STATUS_OKAY()` macros to create a list of `struct i2c_sim_registration` nodes, one for each device.  The `instance` member in each would contain whatever RAM state is necessary as well as a pointer to the data from devicetree in a emulator-specifc `struct config` much like the device-specific `struct config` in each driver implementation.  So each emulator instance is structurally similar to a device---but it's not a device, because the device is the actual `eeprom_at2x` driver.  The emulator instance is just data that's hooked into I2C to allow it to feed the device driver what it would see from hardware.  That data is located by the simulated I2C controller using the address passed to the standard I2C functions.

`emul_atmel_at24_register(i2c)` would invoke `i2c_sim_register(i2c, instances + i)` in a loop over the instances.  (There're details here related to initialization order, but for now that's a start.)

`i2c_sim_register()` would add the registration to a list so it can find it later, and invoke the `init()` function for the instance.

The `i2c_sim` driver could then walk its children and invoke `emul_##DT_CHILD_COMPAT(i)##_register(self)` in its `init` function.  (More magic here to associate different children with the correct bus, but probably not insurmountable.  Make it work with one bus controller instance first.)

That's probably 95% of what we need:
* Each `i2c_sim` device instance is initialized through it's `init()` function like a normal I2C device
* In that init function it invokes `emul_compat_register(self)` for all `compat` that appear in its child devices.
* `emul_compat_register(my_bus)` calls back to `i2c_sim_register(my_bus, &inst)` for all the instances on that bus
* After all devices have been registered (or in `i2c_sim_register`) `emulreg->desc->init(emulreg->instance)` is invoked for all `emulreg` in the registered list.

All we need is to ensure that the `emul_atmel_at24` source module is compiled and linked in to `tests/drivers/eeprom` when it's built for native-posix, so the registration function is visible to the simulator device.  Again start with the emulation layer for `atmel,at24` being in `tests/drivers/eeprom/src/` rather than in a shared area.  I expect there might need to be multiple emulators for a given device depending on what sorts of behavior is to be tested.

With some changes to how device pointers are obtained that may or may not make 2.4 things could get easier, but I stand by my not-yet-proven claim that we don't need anything in the devicetree to make this work except the standard association of i2c devices with an i2c controller.

Is this any more clear?

sjg20 commented:
I think I understand what you are getting at here. But why don't you want the emulators to be devices? They seem to be exactly like devices to me, and it allows us to use the device features of Zephyr to set them up, rather than creating a parallel system.

pabigot commented:
About the only thing I see in common between emulators as I conceptualize them and devices is that both are objects: aggregration of state with behavior accessed through a common API.

But the common API of emulators is not used by anything except the one controller driver that implements a standard Zephyr subsystem API based on behavior that's exposed by the API it uses to talk with the emulators registered with it.  User code is never going to invoke the transfer operation on an I2C device emulator; nor will any other driver implementation.  User code *will* invoke emulator-specific functions on the emulator instance, but that doesn't make the emulator a device: just a pointer that is passed into a globally visible function.  That pointer can be obtained with controller-specific API, such as `void *sim_i2c_get_emulator(struct device *i2c_device, uint16_t addr)`, and passed to emulator-specific API like `sim_eeprom_store_data(struct sim_eeprom *em, off_t addr, const void *ptr, size_t len);`.

However, I'm not trying to dictate the solution to what you have in mind.  I've described the path I would take, which has no requirement to extend the devicetree structure to support an emulator, and doesn't introduce devices that have no software analog in standard applications.  If it makes sense to have the emulators be devices then you should feel free to pursue that path.

sjg20 commented:
Thanks for all the detail. I'm trying your approach to start with but I do have concerns.

An I2C controller may never be directly accessed by user code, .e.g. an I2C EEPROM driver may be used instead. Yet the controller is a driver.

Similarly, an I2C GPIO expander may only be accessed as a GPIO from user code.

To me the test is not whether user code accesses it directly. I think indirect access (i2c emul controller -> at25 emulator) doesn't preclude it being a device.

As you may have guessed, my hope is to make use of this approach more widely than just I2C. Perhaps I should write a short design doc about this once I get v3 up?

pabigot commented:
> An I2C controller may never be directly accessed by user code, .e.g. an I2C EEPROM driver may be used instead. Yet the controller is a driver.

Yes, because it's something that *can* be directly accessed, and that is designed to be accessed from multiple independent clients (e.g. eeproms and IO expanders).   (Also it was already a driver.)   It's not that the common API of emulators is never accessed by user code.   It's that that API is only accessed from one component in a context where that component already must deal with serialized access and internal and external synchronization.
 
> As you may have guessed, my hope is to make use of this approach more widely than just I2C.

I expect that, yes.

> Perhaps I should write a short design doc about this once I get v3 up?

Or before.  Much of Zephyr has a history where the first design artifact is a solution with near-complete functionality, rather than a concept of operations/vision document/plan/requirements.  Perhaps an RFC describing the full vision would be the best place to start.

cfriedt commented:
@sjg20 - I may be approaching this problem from a different starting point, but this is what I quickly threw together for something similar

https://github.com/cfriedt/zephyr/blob/i2c-sim-driver/drivers/i2c/i2c_sim.c
https://github.com/cfriedt/zephyr/blob/i2c-sim-driver/dts/bindings/i2c/zephyr%2Csim-i2c.yaml. 

and then simulating devices via callbacks to their bus address

https://github.com/cfriedt/zephyr/blob/i2c-sim-driver/tests/drivers/i2c/i2c_api/src/sim.c

Are you mostly concerned with I2C peripheral emulation?

cfriedt commented:
It's interesting. In particular, it's nice to specify emulated i2c peripherals in DT. I did something similar [for SPI](https://github.com/cfriedt/zephyr/tree/spi-sim-driver) as well.

I wonder if it would be possible to "recycle" some actual driver code / definitions within the emulated drivers - kind of de-duplicating work.

I wonder if it would it be possible to put the emulated driver in `drivers/sensor/hmc5883l/` as, hmc5883l_emu.c or hmc5883l_sim.c or something? Then it should be possible to include the .h file as well. In some cases, it might also be preferable to include the .c file too (I do that on occasion for nefarious / purposes but that's generally frowned upon, lol).

E.g. in hmc5883l.c
```c
#ifndef DT_DRV_COMPAT
#define DT_DRV_COMPAT honeywell_hmc5883l
#endif

#if DT_DRV_COMPAT == honeywell_hmc5883l
DRIVER_AND_API_INIT(...)
#endif
```
and in hmc5883l_emu.c
```c
#define DT_DRV_COMPAT honeywell_hmc5883l_emu
#include "hmc5883l.c"

// .. other emu code
```

pabigot commented:
@cfriedt Could you provide a case for emulating an I2C device by implementing it as a driver that would hang off the emulated I2C controller, rather than using a real driver and injecting a module to emulate the I2C transactions used by that driver?

At the moment the latter seems significantly more valuable to me; in fact I can't think of a situation where the former is useful.

sjg20 commented:
> > An I2C controller may never be directly accessed by user code, .e.g. an I2C EEPROM driver may be used instead. Yet the controller is a driver.
> 
> Yes, because it's something that _can_ be directly accessed, and that is designed to be accessed from multiple independent clients (e.g. eeproms and IO expanders). (Also it was already a driver.) It's not that the common API of emulators is never accessed by user code. It's that that API is only accessed from one component in a context where that component already must deal with serialized access and internal and external synchronization.

But how about the case where the test wants to adjust the emulator? For example, it might want to tell the emulator to 'go away' from the bus, to check that this is handled correctly. Or it might want to emulate changes in physical hardware, such as the charge level in a battery, so that the reaction of the sensor driver is can be tested. These would normally be handled with direct calls to the emulator:

> 
> > As you may have guessed, my hope is to make use of this approach more widely than just I2C.
> 
> I expect that, yes.
> 
> > Perhaps I should write a short design doc about this once I get v3 up?
> 
> Or before. Much of Zephyr has a history where the first design artifact is a solution with near-complete functionality, rather than a concept of operations/vision document/plan/requirements. Perhaps an RFC describing the full vision would be the best place to start.

Maybe, but honestly I am so confused about the changes you have requested that I'm not sure what I would write. We have some fundamental things to figure out.

For now I'm going to update the PR and see if it is closer to what you want. If we can get to that point, then I'm going to argue for some changes, but at least can do this from the POV of a starting point.

pabigot commented:
> But how about the case where the test wants to adjust the emulator? For example, it might want to tell the emulator to 'go away' from the bus, to check that this is handled correctly. Or it might want to emulate changes in physical hardware, such as the charge level in a battery, so that the reaction of the sensor driver is can be tested. These would normally be handled with direct calls to the emulator:

Exactly.  Assuming the infrastructure I proposed up in https://github.com/zephyrproject-rtos/zephyr/pull/27300#issuecomment-669502021 plus https://github.com/zephyrproject-rtos/zephyr/pull/27300#issuecomment-669502021 that would be something like this:

```
#include "emul_battery.h"

/* There's a batt_dev node label that pulls in a fuel gauge driver that's a child
 * of an i2c controller that's a simulated controller. */
#define NODE DT_NODELABEL(batt_dev)

/* The device from DT_BUS_LABEL(NODE) */
static struct device *i2c;

static void test_charge_change (void)
{
    struct emul_battery *data = sim_i2c_get_emulator(i2c, DT_REG_ADDR(NODE));

    /* Tell the emulator to provide I2C data that indicates 2500 mAh left */
    emul_battery_set_charge(data, 2500);
    // do testing here
    emul_battery_set_charge(data, 5);
    // do testing here
}
```

I'm assuming that you have global functions that manipulate the emulator state, rather than some sort of abstracted channel+value API like that used in the sensor API.  In that case all you need to do is get a reference to the state associated with the emulated instance.

I also assume that there would be crafted emulator modules for different tests, rather than trying to come up with a single emulator module that implements every behavior somebody might want to test.

Does that make things more clear?

sjg20 commented:
> > But how about the case where the test wants to adjust the emulator? For example, it might want to tell the emulator to 'go away' from the bus, to check that this is handled correctly. Or it might want to emulate changes in physical hardware, such as the charge level in a battery, so that the reaction of the sensor driver is can be tested. These would normally be handled with direct calls to the emulator:
> 
> Exactly. Assuming the infrastructure I proposed up in [#27300 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/27300#issuecomment-669502021) plus [#27300 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/27300#issuecomment-669502021) that would be something like this:
> 
> ```
> #include "emul_battery.h"
> 
> /* There's a batt_dev node label that pulls in a fuel gauge driver that's a child
>  * of an i2c controller that's a simulated controller. */
> #define NODE DT_NODELABEL(batt_dev)
> 
> /* The device from DT_BUS_LABEL(NODE) */
> static struct device *i2c;
> 
> static void test_charge_change (void)
> {
>     struct emul_battery *data = sim_i2c_get_emulator(i2c, DT_REG_ADDR(NODE));
> 
>     /* Tell the emulator to provide I2C data that indicates 2500 mAh left */
>     emul_battery_set_charge(data, 2500);
>     // do testing here
>     emul_battery_set_charge(data, 5);
>     // do testing here
> }
> ```
> 
> I'm assuming that you have global functions that manipulate the emulator state, rather than some sort of abstracted channel+value API like that used in the sensor API. In that case all you need to do is get a reference to the state associated with the emulated instance.

Yes that's right. However I still think it makes more sense to use a device here. Here you seem to be assuming that struct emul_battery is common to all battery emulators. But that then means that each one has to have a separate struct for its private data.

> 
> I also assume that there would be crafted emulator modules for different tests, rather than trying to come up with a single emulator module that implements every behavior somebody might want to test.

Yes each device would have its own emulator, unless there is enough commonality to justify doing otherwise.

> 
> Does that make things more clear?

Yes thank you, but I still think it needs work. I've pushed a new version which follows what you suggested, so far as I understand it.

pabigot commented:
> Yes that's right. However I still think it makes more sense to use a device here. Here you seem to be assuming that struct emul_battery is common to all battery emulators.

I'm assuming that `struct emul_battery` is what provides state to the `emul_batt` functions being invoked in that test.  If there's a different emulator it'll have different functions with a different structure.  That's why `sim_i2c_get_emulator()` returns a void pointer: it gets cast to the type appropriate for the emulator registered at the given address.

> But that then means that each one has to have a separate struct for its private data.

Sure, just as every driver has a separate state for its private data.  The difference is the drivers have common API; the emulators don't.  Not even every battery emulator will provide the same functions; possibly several would provide an API to change the charge, but other functions will be specific to whatever the emulator's being used to test.

> Yes thank you, but I still think it needs work. I've pushed a new version which follows what you suggested, so far as I understand it.

I'll take a look, probably tomorrow.

sjg20 commented:
> > Yes that's right. However I still think it makes more sense to use a device here. Here you seem to be assuming that struct emul_battery is common to all battery emulators.
> 
> I'm assuming that `struct emul_battery` is what provides state to the `emul_batt` functions being invoked in that test. If there's a different emulator it'll have different functions with a different structure. That's why `sim_i2c_get_emulator()` returns a void pointer: it gets cast to the type appropriate for the emulator registered at the given address.

I'm not sure it will have different functions. Of course it is possible to do that, but it would be nice to have a common API for each type of device (e.g. sensors, EEPROMs, GPIO expanders).

> 
> > But that then means that each one has to have a separate struct for its private data.
> 
> Sure, just as every driver has a separate state for its private data. The difference is the drivers have common API; the emulators don't. Not even every battery emulator will provide the same functions; possibly several would provide an API to change the charge, but other functions will be specific to whatever the emulator's being used to test.

Perhaps that's the problem we have here. I think each device will need its own emulator, but I am hoping to avoid configuring each one differently. In particular it would be nice to have, for example, an epprom test that can work with all of them.

> 
> > Yes thank you, but I still think it needs work. I've pushed a new version which follows what you suggested, so far as I understand it.
> 
> I'll take a look, probably tomorrow.

Thanks.

- Simon

pabigot commented:
> Perhaps that's the problem we have here. I think each device will need its own emulator, but I am hoping to avoid configuring each one differently. In particular it would be nice to have, for example, an epprom test that can work with all of them.

That may be part of the confusion.  There is exactly one eeprom driver, so yes there should be a single test that will work with it (assuming all the behaviors to be tested were packed into a single emulator).

There are many sensor drivers.  Each would need its own test and matching emulator, and the functions for each could be quite different, certainly in implementation as the I2C protocol would be different for each.  Once you're implementing a custom emulator to support a single test covering the functionality of a single driver, it's not obvios there's value in a generic API.  

I suppose it would be possible to shadow the sensor subsystem channel/value API to make them appear generic. and to have a generic "temperature sensor" test (`sim_sensor_set(emul, SENSOR_CHAN_TEMP, 20.5)`), but I suspect that would have limited value as there's still likely to be driver-specific behavior that it couldn't verify (handling sample checksum violations, or out-of-range values).

sjg20 commented:
> > Perhaps that's the problem we have here. I think each device will need its own emulator, but I am hoping to avoid configuring each one differently. In particular it would be nice to have, for example, an epprom test that can work with all of them.
> 
> That may be part of the confusion. There is exactly one eeprom driver, so yes there should be a single test that will work with it (assuming all the behaviors to be tested were packed into a single emulator).

Well we still want an emulator for each chip I think, so we can actually test that the drivers (such as eeprom_at24x.c) do the right thing. It's just that things are a step removed.

> 
> There are many sensor drivers. Each would need its own test and matching emulator, and the functions for each could be quite different, certainly in implementation as the I2C protocol would be different for each. Once you're implementing a custom emulator to support a single test covering the functionality of a single driver, it's not obvios there's value in a generic API.

So you are strongly arguing against using a device as the abstraction for an emulator?

> 
> I suppose it would be possible to shadow the sensor subsystem channel/value API to make them appear generic. and to have a generic "temperature sensor" test (`sim_sensor_set(emul, SENSOR_CHAN_TEMP, 20.5)`), but I suspect that would have limited value as there's still likely to be driver-specific behavior that it couldn't verify (handling sample checksum violations, or out-of-range values).

Yes I imagine a 'back door' where you can change the value of sensors using an API that could be supported by all sensor emulators. Chargers (for example) are likely to have at least some tests in comment. If we require everything to be different, I don't expect it to scale well.

Another reason is that I'd like to be able to use the device tree to configure the emulators. It is nice to have the config in once place.

cfriedt commented:
> @cfriedt Could you provide a case for emulating an I2C device by implementing it as a driver that would hang off the emulated I2C controller, rather than using a real driver and injecting a module to emulate the I2C transactions used by that driver?
> 
> At the moment the latter seems significantly more valuable to me; in fact I can't think of a situation where the former is useful.

I was only half-awake when I commented. I think my point was just that it would be nice to reuse some amount of code rather than redefining enumerations, defines for register addresses, bit masks, etc

pabigot commented:
> Well we still want an emulator for each chip I think,

Agreed; I hadn't realized there were stm32 and lpc11u6x variants as well, though of course those don't use I2C.  Nor had I realized there was a simulator driver, which like the to-be-removed eeprom_slave driver doesn't warm my heart.  So there's still only one in-tree driver for which an I2C EEPROM emulator is useful.

> So you are strongly arguing against using a device as the abstraction for an emulator?

No, I'm saying I've seen no evidence that it's an appropriate abstraction for emulators as I understand the use case.  For the case of emulating I2C devices I've outlined a solution design that introduces neither devices nor devicetree elements other than the `compatible` property that selects the emulating I2C controller.

> Another reason is that I'd like to be able to use the device tree to configure the emulators. It is nice to have the config in once place.

Maybe.  Generally in Zephyr we restrict devicetree to describing hardware.  Enabling the driver for a particular piece of hardware, and setting options on how that driver behaves, is done via Kconfig.  The prototype hasn't evolved to where there's anything that I could see belonging in devicetree.

pabigot commented on drivers/emul/eeprom_at24_emul.c:
```diff
@@ -0,0 +1,149 @@
+	uint8_t *buf;
+	/** Current register to read */
+	uint8_t cur_reg;
+	/** Current I2C config */
```
Why is this here?  Shouldn't it be retrieved from the I2C controller at runtime?

sjg20 replied:
At present cur_reg is set when you do a byte write to the device.

So you can do:
- write address
- read bytes
as separate I2C transsaction

pabigot commented on drivers/emul/eeprom_at24_emul.c:
```diff
@@ -0,0 +1,149 @@
+	uint16_t size;
+	/** EEPROM data contents */
+	uint8_t *buf;
+	/** Current register to read */
```
If this is current address it needs to be at least 16 bits.

pabigot commented on drivers/i2c/i2c_emul_ctlr.c:
```diff
@@ -0,0 +1,144 @@
+
+	/*
+	 * Tell all the emulators, since they have no way of looking up this
+	 * information otherwise
```
Sure they do.  They know their parent controller, and can get its `struct device*` pointer, and can invoke emulator-specific operations on it.

pabigot reviewed (commented):
I was expecting something that would allow `tests/drivers/eeprom` to pass on native_posix using the AT24 driver rather than the eeprom_simulator driver it currently uses (another thing I didn't realize was in-tree).

There's still a separate hierarchy to provide emulators.  I remain deeply skeptical that a common emulator for any function (like an AT24 EEPROM) is useful for verifying driver implementations.

For that use case, which I believe is the most valuable, I would implement a specific emulator that's part of the test for a hardware driver, with controllable emulated I2C transactions to ensure the test has complete coverage and acceptable behavior for all driver code paths taken when given both normal and abnormal bus transaction behavior.  Such a module has no value outside the driver test area.

I don't see a need for emulators to hook into the device init procedure in order to set them up, nor any other motivation for making them look like devices.

Bottom line: This remains far more complex than I believe is necessary given my understanding of the problem to be solved.

I think other people should step in and provide feedback, because I don't think my perspective is helping us make progress.

Alternatively or concurrently, I'd highly recommend stepping back and providing an RFC issue that details your goals for simulator support in Zephyr, the high level architectural artifacts required to support them, and the use cases that drive the design.  @cfriedt should also participate in that discussion.

sjg20 commented on drivers/emul/eeprom_at24_emul.c:
```diff
@@ -0,0 +1,149 @@
+	uint8_t cur_reg;
+	/** Current I2C config */
+	uint32_t dev_config;
+	/** Address width for EEPROM in bits (8 or 16) */
```
OK. I suppose we can address that when needed

sjg20 commented on drivers/emul/eeprom_at24_emul.c:
```diff
@@ -0,0 +1,149 @@
+	uint32_t dev_config;
+	/** Address width for EEPROM in bits (8 or 16) */
+	uint8_t addr_width;
+	/** true if the bus is too fast */
```
This allows the emulator to fail if the speed is too high for what it expects, so that we can test that behaviour. This is just an example of the sorts of things that can be done.

sjg20 commented:
> I was expecting something that would allow `tests/drivers/eeprom` to pass on native_posix using the AT24 driver rather than the eeprom_simulator driver it currently uses (another thing I didn't realize was in-tree).

That is what I have implemented. We could perhaps drop the eeprom simulator.

> 
> There's still a separate hierarchy to provide emulators. I remain deeply skeptical that a common emulator for any function (like an AT24 EEPROM) is useful for verifying driver implementations.
> 
> For that use case, which I believe is the most valuable, I would implement a specific emulator that's part of the test for a hardware driver, with controllable emulated I2C transactions to ensure the test has complete coverage and acceptable behavior for all driver code paths taken when given both normal and abnormal bus transaction behavior. Such a module has no value outside the driver test area.

I suspect you are thinking that each test is built and run as a separate program. My hope was that we could build a binary with lots of tests in it. Perhaps that doesn't fit well with the Zephyr model?

> 
> I don't see a need for emulators to hook into the device init procedure in order to set them up, nor any other motivation for making them look like devices.

Yes I see you have a much narrower view of the goals. But I really want to get something started, so I can scale this back and perhaps re-architect it later as needed.

> 
> Bottom line: This remains far more complex than I believe is necessary given my understanding of the problem to be solved.
> 
> I think other people should step in and provide feedback, because I don't think my perspective is helping us make progress.
> 
> Alternatively or concurrently, I'd highly recommend stepping back and providing an RFC issue that details your goals for simulator support in Zephyr, the high level architectural artifacts required to support them, and the use cases that drive the design. @cfriedt should also participate in that discussion.

OK, I'll do one more revision to see if it is closer to what you are hoping for. Do you want me to write up my ultimate goal, or the narrower thing that I think you are looking for?

pabigot commented:
> OK, I'll do one more revision to see if it is closer to what you are hoping for. Do you want me to write up my ultimate goal, or the narrower thing that I think you are looking for?

I suggest not bothering with another revision of this; I don't think it has legs in the absence of a plan.

Instead open an RFC issue with your vision of how Zephyr should be supporting emulation in its full depth: how it adds value, where it can be used, the cases where it's not useful.  Once that's described and generally accepted the process toward reaching the goal can be broken down into smaller steps that can be validated against the requirements.  Ignore my perspective except to whatever degree you feel appropriate to address the questions you can expect me to ask about it based on the concerns I've raised here and in the previous attempt.

sjg20 commented:
> > OK, I'll do one more revision to see if it is closer to what you are hoping for. Do you want me to write up my ultimate goal, or the narrower thing that I think you are looking for?
> 
> I suggest not bothering with another revision of this; I don't think it has legs in the absence of a plan.
> 
> Instead open an RFC issue with your vision of how Zephyr should be supporting emulation in its full depth: how it adds value, where it can be used, the cases where it's not useful. Once that's described and generally accepted the process toward reaching the goal can be broken down into smaller steps that can be validated against the requirements. Ignore my perspective except to whatever degree you feel appropriate to address the questions you can expect me to ask about it based on the concerns I've raised here and in the previous attempt.

Oh I didn't see this until now. I just uploaded a very narrow version which I think is what you are looking for. I'll work on the RFC but on consideration I don't think it helps much to do that for the narrower version, so I'll go back to my original goal at least for the first step.

sjg20 commented:
I've had a crack at describing the end goal here: https://github.com/zephyrproject-rtos/zephyr/issues/27531

pabigot commented:
> I just uploaded a very narrow version which I think is what you are looking for. 

Thanks; this is closer.  For my own edification and to validate my understanding I made some changes to bring it closer to what I had proposed.  That's in 3515a476da4953c67eb4d52808fd0f1024c91d13 on branch https://github.com/pabigot/zephyr/commits/npr/27300.

The commit message describes the changes.  This is enough to get a coverage result for the AT24 driver on the standard functional test, which confirms that the success path is pretty well handled, but validation isn't checked nor is use of the option write-protect GPIO.  It, like yours, would have to be split into more self-contained focused commits to be reviewable, and there's some additional refactoring that should be done on the data structures and functions related to associating emulators with the emulating controller.  Also the change replace the test of the simulating driver, which isn't ideal.  And the commit description paragraph suggesting a naming convention is, on further consideration, wrong.

If this is useful you're welcome to merge it or parts of it into your PR, or I can refactor it into chunks that make more sense.

sjg20 committed:
emul: Create an emulation implementation

Create a header file and implementation for emulators. Set up a linker
list so that emulators can be found and initialised at start-up.

Emulators are used to emulate hardware devices, to support testing of
various subsystems. For example, it is possible to write an emulator
for an I2C compass such that it appears on the I2C bus and can be used
just like a real hardware device.

Emulators often implement special features for testing. For example a
compass may support returning bogus data if the I2C bus speed is too
high, or may return invalid measurements if calibration has not yet
been completed. This allows for testing that high-level code can
handle these situations correctly. Test coverage can therefore
approach 100% if all failure conditions are emulated.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: Add an emulator for the Atmel AT24 series

This emulator supports basic reads and writes with 8-bit addressing.
It connects itself to any AT24 devices it finds in the device tree. The
I2C emulation controller driver is used to direct I2C messages from the
AT24 driver to the AT24 emulator.

Signed-off-by: Simon Glass <sjg@chromium.org>
Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

sjg20 committed:
emul: i2c: Add support for I2C emulators

Add an emulation controller which routes I2C traffic to attached
emulators depending on the I2C address selected. This allows drivers
for I2C peripherals to be tested on systems that don't have that
peripheral attached, with the emulator handling the I2C traffic.

Signed-off-by: Simon Glass <sjg@chromium.org>
Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

sjg20 commented:
Hi Peter,

Thanks for your help with this.

I see you actually based it off my previous version which is good because that is a lot closer to what I actually want (a general-purpose emulation thing rather than the more limited single-test approach).

I've tidied this up and bit and pushed it. It still needs a bit of documentation to explain how it works. I can do that assuming this looks good to you.

Regards,
Simon

pabigot commented on include/emul.h:
```diff
@@ -0,0 +1,110 @@
+	const char *label;
+};
+
+/** List of emulators attached to a bus */
```
Why are I2C-specific emulator structures here?  I don't think we'll want to add to this file every time a subsystem is extended with emulator support.

`emul_link` was designed taking only I2C into account.  It's *probably* suitable for other things that are bus children, though, so these could be `emul_bus_child_link` and `emul_bus_child_list`,  if you want to generalize off the one instance.

sjg20 replied:
I've renamed it to refer to a bus.

pabigot commented on include/emul.h:
```diff
@@ -0,0 +1,110 @@
+ * @return 0 if OK
+ * @return negative value on error
+ */
+int emul_init(struct device *dev, const struct i2c_emul_list *list);
```
This also should not be public, certainly not with a generic name like `emul_init` and only supporting `i2c_emul_list` parameters.  It'd be ok as `emul_bus_init()` with the renamed `emul_bus_child_list`.

sjg20 replied:
OK I updated it to refer to a bus

pabigot reviewed (commented):
Two big-ish concerns with this:

* Emulator link structures, emulator init functions, and other emulator-specific material should not be in the generic `<emul.h>` header.  They belong in headers specific to the emulating subsystem, as the mechanism used to associate a device with what supplies emulation for it may vary (and in some cases, like GPIO, the device itself does the emulation).  It's possible we can use a common structure for  "bus-with-children" emulation frameworks, but that won't cover every case.

* We can't change native_posix to use the emulator instead of the simulator, as it would leave the simulator untested.  Some sort of test infrastructure magic will be needed to support both tests.

Other than those points this seems reasonable.  We need to get other people's perspective, though.

sjg20 commented:
> Two big-ish concerns with this:
> 
> * Emulator link structures, emulator init functions, and other emulator-specific material should not be in the generic `<emul.h>` header.  They belong in headers specific to the emulating subsystem, as the mechanism used to associate a device with what supplies emulation for it may vary (and in some cases, like GPIO, the device itself does the emulation).  It's possible we can use a common structure for  "bus-with-children" emulation frameworks, but that won't cover every case.

Yes I missed that. I do want all buses to use a similar appraoch and at the moment I can't see what it wouldn't work. So I have renamed the relevant pieces.

> * We can't change native_posix to use the emulator instead of the simulator, as it would leave the simulator untested.  Some sort of test infrastructure magic will be needed to support both tests.

Yes I've gone back to my original approach of adding a second EEPROM and leaving the first one in place. I've updated the test to run on both.

> 
> Other than those points this seems reasonable. We need to get other people's perspective, though.

OK

sjg20 commented:
Hi Paul, I've updated this based on your comments.

- Simon

sjg20 commented:
How could we get someone else to look at this?

pabigot commented on include/emul.h:
```diff
@@ -0,0 +1,108 @@
+ *
+ * Currently this uses the device node label, but that will go away by 2.5.
+ */
+struct emul_link {
```
This should be `emul_link_for_bus` since it's what gets processed by `emul_list_for_bus`.  We don't know what we'll want for emulator identifiers for things that are not bus children.

sjg20 replied:
done

pabigot commented on include/emul.h:
```diff
@@ -0,0 +1,108 @@
+ * @return 0 if OK
+ * @return negative value on error
+ */
+int emul_init_from_list(struct device *dev,
```
This too is specific to buses and that should be reflected in the name.

sjg20 replied:
done

pabigot commented on drivers/i2c/i2c_emul_ctlr.c:
```diff
@@ -0,0 +1,155 @@
+
+/** Working data for the device */
+struct i2c_emul_data {
+	/* List of i2c_emul_lists associated with the device */
```
name in comment isn't right.

sjg20 replied:
done

pabigot reviewed (commented):
The infrastructure related to the links still isn't consistently limited to bus devices.

pabigot commented:
Added to dev-review to get more eyes on this.  If you want it in 2.4 then add that to the milestones.

sjg20 commented:
Thanks Peter. I added the tag and pushed an update to resolve a conflict.

But I am not sure what to make of this failure with qemu:

https://buildkite.com/zephyr/zephyr/builds/6234#4ff312f9-4ed0-4e55-abd8-bcae80492513

I think I must have a misunderstanding about the execution model, since 'eeprom' is not being set as expected. Do you have any ideas on this please?

pabigot commented:
> But I am not sure what to make of this failure with qemu:
> 
> https://buildkite.com/zephyr/zephyr/builds/6234#4ff312f9-4ed0-4e55-abd8-bcae80492513

Reproducing that locally and looking at the linker map it corresponds to the name eeprom in the bss section.  While you've retained the code that grants the user thread access to the eeprom device, I believe that the pointer being used to provide the device is not itself accessible to user mode.  Thus attempts to read it during the test to get the device to be tested cause a fault.

It's not immediately obvious how to get around this in an application that tests two different drivers.  An approach I took in #27502 was to pull all the test code out into a file that was included twice, once for each context.  Easy there because the language difference supported namespacing, much uglier here.

Another approach would be to implement a [system call](https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html) to fetch the driver pointer from supervisor-accessible data.  That doesn't appear to be trivial, though some mutation of the solution for [driver-specific extensions](https://docs.zephyrproject.org/latest/reference/drivers/index.html#device-specific-api-extensions) should be feasible.

Another is to not try to test both drivers in one application like this, but figure out a way to select the driver via Kconfig and create another test variation that provides the flags that select between the simulated and emulated EEPROM drivers when `native_posix` is used.  Or use `native_posix64` for one of them.

Another is to ask @andrewboie if there's a way to make the pointer to the device accessible to user mode.

Those are probably in increasing order of simplicity, so I'd start with the last.

andrewboie commented:
> Another is to ask @andrewboie if there's a way to make the pointer to the device accessible to user mode.

`device_get_binding()` is a system call.

pabigot commented:
> `device_get_binding()` is a system call.

Yes, but I think that just adds a level of indirection.  How would we communicate the name of the device into each test so it could use `device_get_binding(devname)` like it did originally (when `devname` was a string constant).

andrewboie commented:
unless you are asking to make the `struct device` data itself accessible to user mode, and not just its memory address? we don't allow this for security reasons

andrewboie commented:
> Yes, but I think that just adds a level of indirection. How would we communicate the name of the device into each test so it could use `device_get_binding(devname)` like it did originally (when `devname` was a string constant).

Hmm I might not be understanding the problem.

The actual name of the device to look up is usually some #defined string, no?

pabigot commented:
> The actual name of the device to look up is usually some #defined string, no?

Usually, yes.  In this case we want to run the same test code on two different devices, so it's no longer a constant.

I think this points to the diff: https://github.com/zephyrproject-rtos/zephyr/pull/27300/files#diff-511e7ff7c3ee95bd9e4671a72f5fd6f9

andrewboie commented:
Ohhh, easy fix for this one. What's happening is that you have a global variable, but need to add it to ztest's memory domain otherwise user mode can't use it.

Change

```
static struct device *eeprom;
```

to

```
ZTEST_BMEM static struct device *eeprom;
```

and I bet it works. `ZTEST_BMEM` is just a macro for `#define ZTEST_BMEM  K_APP_BMEM(ztest_mem_partition)`

Background reading:

https://docs.zephyrproject.org/latest/reference/usermode/memory_domain.html#memory-domains

pabigot commented:
@andrewboie Thanks, that does work.

pabigot commented:
@sjg20 I think you should use that technique to instead provide the name of the device through the global, then revert the changes so the individual tests still do the `device_get_binding()` themselves.  Partly to reduce the changes from the original, but also so that diagnostics and status messages in the tests can display the identity of the device instance being tested.  (Such diagnostics would have to be added as the original assumed a single device for the whole test.)

sjg20 commented:
> @sjg20 I think you should use that technique to instead provide the name of the device through the global, then revert the changes so the individual tests still do the `device_get_binding()` themselves. Partly to reduce the changes from the original, but also so that diagnostics and status messages in the tests can display the identity of the device instance being tested. (Such diagnostics would have to be added as the original assumed a single device for the whole test.)

Done, thank you both.

pabigot reviewed (approved):
I would still like to see more systems engineering and coordination on what we want a Zephyr emulator framework to be, but that isn't going to happen.

It should be understood that all the emulator stuff is "experimental" in the [API sense](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#experimental) even though I don't think "emulators" as a category do or should define a generic API.

Given this expectation I think it's worth getting this in so people can start seeing whether it's useful.  I need the GPIO emulator from #26484 before I can do much with an I2C emulation framework.

pabigot commented:
BTW: The buildkite failure is #27847.  Odd we haven't seen that before.

mbolivar-nordic commented:
Dev-review: this was brought up; @pabigot said his above comment had everything he wanted to say on the matter. No other comments.

sjg20 commented:
https://github.com/zephyrproject-rtos/zephyr/pull/27848 fixes the buildkite error, so once that lands I'll update this.

nashif commented on include/emul.h:
```diff
@@ -0,0 +1,108 @@
+/*
```
This file should probably be in include/drivers/emul.h given that the implementation is in drivers/emul

nashif replied:
> so maybe the implementation should move

yes, move it to subsys/emul maybe

pabigot replied:
Yes, that  makes sense.

@sjg20 could you move the files from `drivers/emul/*` to `subsys/emul/i2c/` and change `eeprom_at24_emul.c` to `emul_atmel_at24.c`?  Rationale:
* This driver is specific to the i2c emulator API so should be a directory distinct from any future emulator APIs
* Globally available emulators should use the compatible name (`atmel,at24` in this case) as the pattern for their primary identifier.  The Kconfig prefix should match: `EMUL_I2C_ATMEL_AT24`.

I don't think it's necessary to make big changes to `drivers/i2c_emul` because that actually is an I2C driver for emulators.  It's in the right place, but for consistency with the compatible and header the implementation file would be better named `i2c_emul.c`.

sjg20 commented:
The emul.c and emul.h files are not actually specific to i2c. I expect to use them for SPI also, at least. I don't mind moving it into subsys/emul/i2c/ for now, if it helps get us a starting point, then moving back to subsys/emul later?

pabigot commented:
> The emul.c and emul.h files are not actually specific to i2c. I expect to use them for SPI also, at least. I don't mind moving it into subsys/emul/i2c/ for now, if it helps get us a starting point, then moving back to subsys/emul later?

Right, `emul.c` and `Kconfig` belong into `subsys/emul`.  It's the eeprom one which is I2C-specific and should go into `subsys/emul/i2c`

sjg20 commented:
Does this look better?
